basics/create-a-solver/create-a-solver.jmd

```diff
@@ -0,0 +1,205 @@
+contains a lot of useful packages, we'll be using a collection of those to make
+our comparison easier. The most important one by far is [SolverBenchmarks](https://github.com/JuliaSmoothOptimizers/SolverBenchmark.jl), a
+benchmarking package designed for manipulating DataFrames that contain statistics
+regarding the execution of a specific sover given a set of problems.
+
+Let us compare our ```newton``` method to a solver contained in the
+[JSOSolvers](https://github.com/JuliaSmoothOptimizers/JSOSolvers.jl) package, namely the ```lbfgs``` one.
+Firstly, we'll need to gather our data regarding each solver.
+```julia
+using SolverBenchmark, JSOSolvers
+solvers = Dict(:newton=>newton, :lbfgs=>lbfgs)
+f1(x) = x[1]^2 + x[2]^2
+f2(x) = (1 - x[1])^2 + 100(x[2] - x[1]^2)^2
+f3(x) = (x[1]^2 + x[2] - 11) + (x[1] + x[2]^2 - 7)^2
+f4(x) = 0.26(x[1]^2 + x[2]^2) - 0.48*x[1]*x[2]
+test_functions = [f1, f2, f3, f4]
+problems = (ADNLPModel(i, 2*ones(2),name="Problem $i") for i in test_functions)
+stats = bmark_solvers(solvers, problems)
+```
+The ``bmark_solvers`` function is also part of the **SolverTools** package, and
+runs our solver on a set of problems, in our case ``f1``, ``f2``, ``f3`` and ``f4``.
+Lastly, it returns a DataFrame of statistics that we can manipulate using
+**SolverBenchmarks**.
+
+Next, let's create a latex table with the data we just gathered
+```julia
+open("newton.tex","w") do io
+	latex_table(io,stats[:newton],cols = [:name,:status,:objective,:elapsed_time,:iter])
+end
+open("lbfgs.tex","w") do io
+	latex_table(io,stats[:lbfgs],cols = [:name,:status,:objective,:elapsed_time,:iter])
+end
+
+```
+
+That will give us a nicely formatted table that we can just plug into our
+latex code. It's also possible to create the table in the markdown format.
+
+```julia
+open("newton.md","w") do io
+	markdown_table(io,stats[:newton],cols = [:name,:status,:objective,:elapsed_time,:iter])
+end
+open("lbfgs.md","w") do io
+	markdown_table(io,stats[:lbfgs],cols = [:name,:status,:objective,:elapsed_time,:iter])
+end
+```
+Either way, our tables should look something like this:
+### LBFGS
+
+```
+|       name |      status | objective | elapsed_time | iter |
+|------------|-------------|-----------|--------------|------|
+| Problem f1 | first_order |   0.0e+00 |      1.2e+00 |    2 |
+| Problem f2 | first_order |   1.3e-16 |      7.0e-04 |   25 |
+| Problem f3 | first_order |  -8.4e+00 |      3.7e-04 |    8 |
+| Problem f4 | first_order |   1.2e-28 |      9.8e-05 |    2 |
+```
+### Newton
+
+```
+|       name |      status | objective | elapsed_time | iter |
+|------------|-------------|-----------|--------------|------|
+| Problem f1 | first_order |   0.0e+00 |      6.1e-01 |    2 |
+| Problem f2 | first_order |   0.0e+00 |      1.5e-01 |   17 |
+| Problem f3 | first_order |  -8.4e+00 |      1.5e-01 |   16 |
+| Problem f4 | first_order |   5.6e-62 |      1.4e-01 |    2 |
+```
+In the latex format
+
+![](lbfgs.svg)
+
+![](newton.svg)
```

Comment:
Since we're auto-building, we may have to remove this, or else create it programmatically.